Core/Core/Use Cases/UploadManager.swift

```diff
@@ -215,32 +198,81 @@ open class UploadManager: NSObject, URLSessionDelegate, URLSessionTaskDelegate,
     }
 
     public func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
-        Logger.shared.log()
         context.performAndWait {
             guard
                 let step = task.uploadStep,
                 let file = self.file(taskID: task.taskIdentifier)
             else { return }
+            let decoder = JSONDecoder()
+            decoder.dateDecodingStrategy = .iso8601
             switch step {
             case .upload:
-                if case let .submission(courseID, assignmentID, comment)? = file.context, error == nil {
-                    submit(file: file, courseID: courseID, assignmentID: assignmentID, comment: comment)
-                    return
+                // On device the upload might fail with a 201
```

Comment:
I'm not too familiar with File Uploads so take this with a grain of salt.  This method seems very large and to have a ton of different logic in it.  It appears you tested it well. Would it be helpful or possible (maybe it's not helpful or possible :shrug:) to break this into smaller chunk size methods and easier to map the testing to individual methods?  Just a thought.

Reply:
I agree it's pretty difficult to follow. I will break it up.

Reply:
I'm also going to see if the `fail with a 201` is still a thing. It only happened on Bitrise build so I'll have to wait for it to build.

Reply:
I pushed up a refactor. I think I still prefer always testing through `urlSession(_:task:didCompleteWithError:)` since that is the entry point but I already like having that function split up and chunked.

Thanks for the feedback.